File: app/golang/internal/order/domain/order.go
package domain

import (
	"github.com/google/uuid"
	productDomain "github.com/orkhan/ecommerce/internal/product/domain"
	sharedDomain "github.com/orkhan/ecommerce/internal/shared/domain"
)

// Order — Aggregate Root
//
// Laravel: Order.php  ·  Spring: Order.java
type Order struct {
	sharedDomain.AggregateRoot

	id          OrderID
	userID      uuid.UUID
	items       []OrderItem
	status      OrderStatus
	totalAmount productDomain.Money
	address     Address
}

func Create(userID uuid.UUID, items []OrderItem, address Address, currency productDomain.Currency) (*Order, error) {
	if len(items) == 0 {
		return nil, sharedDomain.NewDomainError("Sifariş ən azı 1 məhsuldan ibarət olmalıdır")
	}

	id := GenerateOrderID()
	total := productDomain.ZeroMoney(currency)
	for _, item := range items {
		next, err := total.Add(item.LineTotal())
		if err != nil {
			return nil, err
		}
		total = next
	}

	o := &Order{
		id:          id,
		userID:      userID,
		items:       items,
		status:      OrderStatusPending,
		totalAmount: total,
		address:     address,
	}

	o.RecordEvent(OrderCreatedEvent{
		BaseEvent: sharedDomain.NewBaseEvent(),
		OrderID:   id,
		UserID:    userID,
	})
	o.RecordEvent(OrderCreatedIntegrationEvent{
		BaseEvent:   sharedDomain.NewBaseEvent(),
		OrderID:     id.UUID(),
		UserID:      userID,
		TotalAmount: total.Amount(),
		Currency:    string(currency),
	})
	return o, nil
}

func Reconstitute(id OrderID, userID uuid.UUID, items []OrderItem, address Address,
	status OrderStatus, totalAmount productDomain.Money) *Order {
	return &Order{
		id: id, userID: userID, items: items, address: address,
		status: status, totalAmount: totalAmount,
	}
}

func (o *Order) Confirm() error {
	if err := o.status.RequireTransitionTo(OrderStatusConfirmed); err != nil {
		return err
	}
	o.status = OrderStatusConfirmed
	o.RecordEvent(OrderConfirmedEvent{BaseEvent: sharedDomain.NewBaseEvent(), OrderID: o.id})
	return nil
}

func (o *Order) MarkAsPaid() error {
	if err := o.status.RequireTransitionTo(OrderStatusPaid); err != nil {
		return err
	}
	o.status = OrderStatusPaid
	o.RecordEvent(OrderPaidEvent{BaseEvent: sharedDomain.NewBaseEvent(), OrderID: o.id})
	return nil
}

func (o *Order) Ship() error {
	if err := o.status.RequireTransitionTo(OrderStatusShipped); err != nil {
		return err
	}
	o.status = OrderStatusShipped
	return nil
}

func (o *Order) Deliver() error {
	if err := o.status.RequireTransitionTo(OrderStatusDelivered); err != nil {
		return err
	}
	o.status = OrderStatusDelivered
	return nil
}

func (o *Order) Cancel(reason string) error {
	if o.status == OrderStatusDelivered {
		return sharedDomain.NewDomainError("Çatdırılmış sifarişi ləğv etmək olmaz")
	}
	if o.status == OrderStatusCancelled {
		return sharedDomain.NewDomainError("Sifariş artıq ləğv edilib")
	}
	o.status = OrderStatusCancelled
	o.RecordEvent(OrderCancelledEvent{
		BaseEvent: sharedDomain.NewBaseEvent(), OrderID: o.id, Reason: reason,
	})
	return nil
}

func (o *Order) ID() OrderID                       { return o.id }
func (o *Order) UserID() uuid.UUID                 { return o.userID }
func (o *Order) Items() []OrderItem                { return o.items }
func (o *Order) Status() OrderStatus               { return o.status }
func (o *Order) TotalAmount() productDomain.Money  { return o.totalAmount }
func (o *Order) Address() Address                  { return o.address }

// === REPOSITORY ===
type Repository interface {
	Save(order *Order) error
	FindByID(id OrderID) (*Order, error)
	FindByUserID(userID uuid.UUID) ([]*Order, error)
}

// === SPECIFICATION ===
func OrderCanBeCancelled() sharedDomain.Specification[*Order] {
	return sharedDomain.SpecFunc[*Order](func(o *Order) bool {
		return o.status == OrderStatusPending || o.status == OrderStatusConfirmed
	})
}

File: app/spring/src/main/java/az/ecommerce/order/domain/Order.java
package az.ecommerce.order.domain;

import az.ecommerce.order.domain.enums.OrderStatusEnum;
import az.ecommerce.order.domain.event.*;
import az.ecommerce.order.domain.valueobject.Address;
import az.ecommerce.order.domain.valueobject.OrderId;
import az.ecommerce.order.domain.valueobject.OrderItem;
import az.ecommerce.product.domain.valueobject.Currency;
import az.ecommerce.product.domain.valueobject.Money;
import az.ecommerce.shared.domain.AggregateRoot;
import az.ecommerce.shared.domain.exception.DomainException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * === ORDER AGGREGATE ROOT ===
 *
 * Laravel: src/Order/Domain/Entities/Order.php
 *
 * Bu, layihənin ən kompleks aggregate-idir:
 * - State machine (OrderStatusEnum)
 * - Domain events
 * - Integration events
 * - Embedded VO-lar (Address, Money)
 * - Entity collection (OrderItem-lər)
 */
public class Order extends AggregateRoot {

    private final OrderId id;
    private final UUID userId;
    private final List<OrderItem> items;
    private OrderStatusEnum status;
    private Money totalAmount;
    private final Address address;

    private Order(OrderId id, UUID userId, List<OrderItem> items, Address address, Currency currency) {
        if (items.isEmpty()) {
            throw new DomainException("Sifariş ən azı 1 məhsuldan ibarət olmalıdır");
        }
        this.id = id;
        this.userId = userId;
        this.items = new ArrayList<>(items);
        this.address = address;
        this.status = OrderStatusEnum.PENDING;
        this.totalAmount = calculateTotal(items, currency);
    }

    public static Order create(UUID userId, List<OrderItem> items, Address address, Currency currency) {
        OrderId id = OrderId.generate();
        Order order = new Order(id, userId, items, address, currency);
        order.recordEvent(OrderCreatedEvent.of(id, userId, order.totalAmount.amount(), currency, items.size()));
        order.recordEvent(new OrderCreatedIntegrationEvent(
                UUID.randomUUID(), java.time.Instant.now(),
                id.value(), userId, order.totalAmount.amount(), currency.name()));
        return order;
    }

    public static Order reconstitute(OrderId id, UUID userId, List<OrderItem> items,
                                     Address address, OrderStatusEnum status, Money totalAmount) {
        Order order = new Order(id, userId, items, address, totalAmount.currency());
        order.status = status;
        order.totalAmount = totalAmount;
        return order;
    }

    public void confirm() {
        status.requireTransitionTo(OrderStatusEnum.CONFIRMED);
        this.status = OrderStatusEnum.CONFIRMED;
        recordEvent(OrderConfirmedEvent.of(id));
    }

    public void markAsPaid() {
        status.requireTransitionTo(OrderStatusEnum.PAID);
        this.status = OrderStatusEnum.PAID;
        recordEvent(OrderPaidEvent.of(id));
    }

    public void ship() {
        status.requireTransitionTo(OrderStatusEnum.SHIPPED);
        this.status = OrderStatusEnum.SHIPPED;
        recordEvent(OrderShippedEvent.of(id));
    }

    public void deliver() {
        status.requireTransitionTo(OrderStatusEnum.DELIVERED);
        this.status = OrderStatusEnum.DELIVERED;
        recordEvent(OrderDeliveredEvent.of(id));
    }

    public void cancel(String reason) {
        if (status == OrderStatusEnum.DELIVERED) {
            throw new DomainException("Çatdırılmış sifarişi ləğv etmək olmaz");
        }
        if (status == OrderStatusEnum.CANCELLED) {
            throw new DomainException("Sifariş artıq ləğv edilib");
        }
        this.status = OrderStatusEnum.CANCELLED;
        recordEvent(OrderCancelledEvent.of(id, reason));
    }

    private Money calculateTotal(List<OrderItem> items, Currency currency) {
        Money total = Money.zero(currency);
        for (OrderItem item : items) {
            total = total.add(item.lineTotal());
        }
        return total;
    }

    public OrderId id() { return id; }
    public UUID userId() { return userId; }
    public List<OrderItem> items() { return Collections.unmodifiableList(items); }
    public OrderStatusEnum status() { return status; }
    public Money totalAmount() { return totalAmount; }
    public Address address() { return address; }
}
